File: Game of Subsets[25-April-2023]/Solution.cpp
//   >> C++ CODE <<


class Solution {
    int mod;
    vector<int> mp;
    public:
        Solution() {
            mod = (int) 1e9 + 7;
            mp.assign(31, 0);
            int prime[] = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29};
            for (int i = 2; i <= 30; ++i) {
                if (i % 4 == 0 || i % 9 == 0 || i == 25) continue;
                int mask = 0;
                for (int j = 0; j < 10; ++j) {
                    if (i % prime[j] == 0) mask |= 1 << j;
                }
                mp[i] = mask;
            }
        }
    
        long long pow(int n) {
            long long ans = 1, m = 2;
            while (n != 0) {
                if ((n & 1) == 1) ans = (ans * m) % mod;
                m = (m * m) % mod;
                n >>= 1;
            }
            return ans;
        }
    
        int goodSubsets(vector<int>& arr, int n) {
            int one = 0;
            vector<int> dp(1024), cnt(31);
            dp[0] = 1;
            for (int i = 0; i < n; ++i) {
                if (arr[i] == 1) {
                    one++;
                } else if (mp[arr[i]] != 0) {
                    cnt[arr[i]]++;
                }
            }
            for (int i = 0; i < 31; ++i) {
                if (cnt[i] == 0) continue;
                for (int j = 0; j < 1024; ++j) {
                    if ((j & mp[i]) != 0) continue;
                    dp[j | mp[i]] = (dp[j | mp[i]] + dp[j] * 1LL * cnt[i]) % mod;
                }
            }
            long long ans = 0;
            for (int i = 0; i < 1024; ++i) {
                ans = (ans + dp[i]) % mod;
            }
            ans--;
            if (one != 0) {
                ans = (ans * pow(one)) % mod;
            }
            return (int) ans;
        }
};

File: Game of Subsets[25-April-2023]/solution.py
#   >> Python code <<

from typing import List


class Solution:
    def __init__(self):
        self.mod = int(1e9 + 7)
        self.mp = [0]*31
        prime = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
        for i in range(2, 31):
            if i % 4 == 0 or i % 9 == 0 or i == 25:
                continue
            mask = 0
            for j in range(10):
                if i % prime[j] == 0:
                    mask |= 1 << j
            self.mp[i] = mask
    
    def pow(self, n):
        ans, m = 1, 2
        while n != 0:
            if n & 1 == 1:
                ans = (ans * m) % self.mod
            m = (m * m) % self.mod
            n >>= 1
        return ans
    
    def goodSubsets(self,n, arr):
        one = 0
        dp = [0]*1024
        cnt = [0]*31
        dp[0] = 1
        for i in arr:
            if i == 1:
                one += 1
            elif self.mp[i] != 0:
                cnt[i] += 1
        for i in range(31):
            if cnt[i] == 0:
                continue
            for j in range(1024):
                if (j & self.mp[i]) != 0:
                    continue
                dp[j | self.mp[i]] = (dp[j | self.mp[i]] + dp[j] * cnt[i]) % self.mod
        ans = 0
        for i in dp:
            ans = (ans + i) % self.mod
        ans -= 1
        if one != 0:
            ans = (ans * self.pow(one)) % self.mod
        return ans